src/vm/object_model.rs

```diff
@@ -84,6 +84,25 @@ pub trait ObjectModel<VM: VMBinding> {
     /// The metadata specification for the mark-and-nursery bits, used by most plans that has large object allocation. 2 bits.
     const LOCAL_LOS_MARK_NURSERY_SPEC: VMLocalLOSMarkNurserySpec;
 
+    /// The remainder of dividing the address of any valid non-null object reference by the region
+    /// size (in bytes) of the `ALLOC_BIT` side meta data.
+    ///
+    /// The region size is conveniently provided as `crate::vm::ALLOC_BIT_REGION_SIZE`;
+    ///
+    /// This needs to be a constant for any VM that needs to use the `is_mmtk_object` feature.
+    /// Otherwise, the `ALLOC_BIT` side meta data cannot decide which byte in the region contains
+    /// the valid object reference even if the bit for that region is `1`.
+    ///
+    /// If a valid object references *always points to the start of an objects*, the value of this
+    /// constant should be **zero**.
+    ///
+    /// If the offset from the start of an object to its object reference is *always a multiple of
+    /// the region size*, the value of this constant should also be **zero**. The region size of
+    /// `ALLOC_BIT` is currently set to the minimum object alignment (4 bytes on 32-bit, and 8
+    /// bytes on 64-bit).
+    #[cfg(feature = "is_mmtk_object")]
+    const OBJREF_ALLOC_BIT_REM: usize;
```

Comment:
This is a very strong requirement. This essentially requires that all the object references starts at the same byte in a word. It is very likely to be true for a binding, but we cannot guarantee that's always true for all the VMs.

For 64 bits, our default min alignment is 4 bytes and min object size is 8 bytes. That means a binding may get addresses like `0x0`, `0x14` from allocation. If the binding uses the addresses as object refs, those two object refs do not have the same `OBJREF_ALLOC_BIT_REM`.

I think this should be left to the binding to check. That would save us lots of trouble to make sure it is general and correct, and for each specific binding, it is very easy for them to filter out obviously invalid references based on the information they have.

Reply:
The alignment requirement is `addr % MIN_ALIGN == OBJREF_ALLOC_BIT_REM`. It is `% MIN_ALIGN`, not `% MIN_SIZE`. In this case, `0x0` and `0x14` are both congruent to 0 modulo 4 (not modulo 8).

The requirement that "all valid objrefs must have the same remainder" is strong, but I think it has to be this strong.  

Remember that [JikesRVM finds the header size of a valid objref by loading from the object header](https://github.com/mmtk/jikesrvm/blob/master/rvm/src/org/jikesrvm/objectmodel/JavaHeader.java#L328).  When we are given an address to check if it is a valid ObjectReference, it can be an arbitrary address.  It is not safe to load from an arbitrary address, not even when it is in one of MMTk spaces.

`addr % MIN_ALIGN` can be 0, 1, 2, or 3 if `MIN_ALIGN == 4`.  If different objrefs may have different remainders, it is impossible to tell which is the proper remainder.  We can't look at object headers because we can't load from arbitrary addresses.  Then we can only find the proper remainder from other data sources, such as address ranges and side metadata.  But I don't think either of them is practical.

But anyway, we can offload this to the VM, because in practice, the VM will do the pre-check anyway.